better handle boot entry names

TobiPeterG opened the pull request:
This forces that any old boot entries for the same snapshot are removed, e.g. after enabling/diabling boot counting. This will also make sure to reset boot counting after making a change.

I already tested it and it seems to work perfectly :)

Solves #353

TobiPeterG commented:
> What is the issue that is fixing? Is like disabling the boot assessment counter while there are entries with the counter in the name?

More or less the same as before: when an entry without boot counting exists, sdbootutil just creates a new one with boot counting without removing the old one without boot counting (e.g. because the boot was successful it was removed). Also, I think when an entry has e.g. +2-1, the old version will just reuse the name instead of resetting boot counting. 
I also think that when you disable boot counting, it won't remove the old one with boot counting still enabled. 

This new version will check if an entry already exists, if yes, compare
-> no change -> don't write a new one
-> change -> write a new one with the expected entry name -> check if any other entry for this snapshot exists with a different name -> remove those

This should handle all cases correctly, but please let me know if I missed something. 
I'll create a new issue showing the issue I have with the current version in Tumbleweed. :)

aplanas commented on sdbootutil:
```diff
@@ -1153,13 +1153,45 @@ install_kernel()
 		fi
+		target_entry="$boot_root/loader/entries/$target_name"
+
+		if [ -e "$target_entry" ]; then
```
> It is relevant if it changed or not: because if the content didn't change, we do not want to reset boot counting for it, only if it did.

Ah this makes sense, yes.  You mean that when the same initrd is generated, then we do not want to resent the boot counter.

TobiPeterG commented on sdbootutil:
```diff
@@ -1153,13 +1153,45 @@ install_kernel()
+			install_with_rollback "$tmpdir/entry.conf" "$target_entry" || failed="bootloader entry"
+		fi
+
+		# Remove any other variant(s) (with or without boot counting)
```
> I think that this is dangerous. In #356 the boot entry gets removed and the new one with the full name is put in place. There is a moment that there is no boot entry, so the kernel is not reachable, but this is kind of OK as when a new kernel gets installed the last step is the boot entry creation, meaning that this kernel is always unbootable until the correct boot entry is in place.
> 

Are you sure about this? I think the old initrd won't be removed until bootctl cleanup is called further down, so the kernel should be as bootable as it was with the old entry. So yes, there is a moment that no boot entry can boot the kernel, but I don't see the upside of that approach

> With this code we can have two boot entries, one correct and another one pointing to an initrd that can or cannot be installed. I think that #356 is wrong, as should remove the entry early, and gives the old initrd the chance to be garbage collected, but this code can have a duplication. Or am I missing something?

This is correct. I guess we should do something similar as in the install_with_rollback function where the "old" entry is renamed to .bak to not have two entries for the same snapshot.

Independent of that change, the old initrd will be garbage collected by bootctl cleanup below if no other boot entry points to it or am I missing something?

TobiPeterG replied:
I do it similar to the install_with_rollback function handles these cases now, so I think this is resolved :)

TobiPeterG commented:
@aplanas I reworked the logic. It's now assumed that at maximum one entry already exists for the current snapshot/kernel combination. In theory, we already know that this is not true as systems affected by this bug will have two. However, it makes the logic cleaner, should we "fix" the systems with multiple entries for the same kernel/snapshot (which one to keep?) or not?

sdbootutil now also backs up the old entry like install_with_rollback does when we install a new entry with a different name and clean up afterwards if everything suceeds (or restore the old entry and remove the new one if something failed).
I tested it on Tumbleweed with the following cases successfully:
a) old entry exists with(out) boot counting, new changed one is created with boot counting (resets boot counting)
b) old entry exists with(out) boot counting, new equal one is created with boot counting (does not reset boot counting)
c) old entry exists with boot counting, new equal one is created with boot counting disabled (does not reset boot counting)
d) old entry exists with boot counting, new changed one is created with boot counting disabled (removes boot counting)
e) old entry does not exist (new one is created)
f) old entry exists without boot counting, new equal one is created with boot counting disabled (reuses existing one)
g) old entry exists without boot counting, new changed one is created with boot counting disabled (just replaces the file)

I think this should cover all cases

NOTE: The current implementation will _not_ update the tries when e.g. boot counting has been disabled or the boot attempts have been changed, when the content of the entries doesn't change. If it was set to +3 and changed to +4, but the entries stay the same, they will stay with +3. I feel like we should detect these cases where entries have boot counting and we detect a change (e.g. has been disabled or value changed). What do you think? I will create a protoype for how that could look like.

EDIT: I will just add a flag to indicate that the boot counting should be reset to the newest value.

TobiPeterG commented:
Done, works :)

TobiPeterG commented on sdbootutil:
```diff
@@ -1153,13 +1153,45 @@ install_kernel()
+		if [ -n "$tries" ]; then
+			target_name="$(entry_conf_file "$kernel_version" "$snapshot" "$tries")"
+		else
+			target_name="$entry_name"
```
Done
